src/metabase/models/query_field.clj

```diff
@@ -1,10 +1,56 @@
 (ns metabase.models.query-field
   (:require
+   [metabase.lib.util.match :as lib.util.match]
+   [metabase.native-query-analyzer :as query-analyzer]
+   [metabase.util.log :as log]
    [methodical.core :as methodical]
    [toucan2.core :as t2]))
 
-
 (methodical/defmethod t2/table-name :model/QueryField [_model] :query_field)
 
 (doto :model/QueryField
   (derive :metabase/model))
+
+;;; Updating QueryField from card
+
+(defn- field-ids-for-mbql
+  "Find out ids of all fields used in a query. Conforms to the same protocol as [[query-analyzer/field-ids-for-sql]],
+  so returns `{:direct (...int ids)}` map.
+
+  Does not track wildcards for queries rendered as tables afterwards."
+  [query]
+  {:direct (some-> (lib.util.match/match query [:field (id :guard integer?) _] id)
```

Comment:
there is `metabase.lib.util/referenced-field-ids` that do exactly this

Reply:
Actually, I tested this and `metabase.lib.util/referenced-field-ids` doesn't even get all the implicit joined fields... that's a bug I guess.

Reply:
You could fix `metabase.lib.util/referenced-field-ids` in the same way you did for the migration, getting the `:source-field` out from `opts` if present.

Reply:
oh nice find. True, you could have used that. Now that you've fixed `metabase.lib.util/referenced-field-ids` it doesn't have test coverage. Maybe you could add a test in `parse-mbql-test`?

Reply:
Alternatively, we could add a test in Ngoc's `pmbql->field-usages-test` which also uses that function